{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "prescription-token",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mThe `size` argument of `element_rect()` is deprecated as of ggplot2 3.4.0.\n",
      "\u001b[36mℹ\u001b[39m Please use the `linewidth` argument instead.”\n"
     ]
    }
   ],
   "source": [
    "project_dir = \"~/ben_lustre/current_projects/kidney_glomTI_response\"\n",
    "setwd(project_dir)\n",
    "suppressPackageStartupMessages(source(\"https://raw.githubusercontent.com/bjstewart1/helpful_singlecell/main/functions.R\"))\n",
    "ad <- import(\"anndata\")\n",
    "sc <- import(\"scanpy\")\n",
    "scv <- import(\"scvelo\")\n",
    "sp = import(\"scipy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "upper-running",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "python:         /home/jovyan/my-conda-envs/myenv/bin/python\n",
       "libpython:      /home/jovyan/my-conda-envs/myenv/lib/libpython3.8.so\n",
       "pythonhome:     /home/jovyan/my-conda-envs/myenv:/home/jovyan/my-conda-envs/myenv\n",
       "version:        3.8.8 | packaged by conda-forge | (default, Feb 20 2021, 16:22:27)  [GCC 9.3.0]\n",
       "numpy:          /home/jovyan/my-conda-envs/myenv/lib/python3.8/site-packages/numpy\n",
       "numpy_version:  1.20.3\n",
       "\n",
       "NOTE: Python version was forced by RETICULATE_PYTHON"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "buried-lodging",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "tenx_dir <- \"./data/seq_data/10X_3p\"\n",
    "sequencing_inventory <- read.csv(\"/nfs/team297/bs16/sample_donor_data/sequencing_inventory.csv\", check.names = FALSE)\n",
    "sequencing_inventory <- sequencing_inventory[sequencing_inventory$experiment %in% c(\"10X_3'_V3\") & #just 3p\n",
    "                                           sequencing_inventory$cellranger_SOC %in% 'TRUE' & sequencing_inventory$poor_QC %in% \"FALSE\",]\n",
    "#exclude fetal - these are one per channel and we will analyse these elsewhere\n",
    "sequencing_inventory <- sequencing_inventory[!sequencing_inventory$tissue %in% grep(\"fetal\", sequencing_inventory$tissue, value = TRUE), ]\n",
    "#remove perfusion samples\n",
    "perfusion_batches = c(\"Batch_9\", \"Batch_12\", \"Batch_14\")\n",
    "sequencing_inventory = sequencing_inventory[!sequencing_inventory$tissue_dissociation_processing_batch %in% perfusion_batches, ]\n",
    "#fix the irods spaces\n",
    "sequencing_inventory$irods_ID <- gsub(\" \", \"\", sequencing_inventory$irods_ID)\n",
    "\n",
    "#make a set of channels and files\n",
    "fn_in <- data.frame(\"channel\" = sequencing_inventory$sanger_ID, \"filename\" = sequencing_inventory$irods_ID)\n",
    "\n",
    "#now get the assignments\n",
    "cell_assignments <- readRDS(file.path(project_dir,  \"data/genotypes/cell_assignments_cells.RDS\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "synthetic-accuracy",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***====================***\n",
      "\n",
      "2.08333333333333%\n",
      "\n",
      "sample CG_SB_NB8768284\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample CG_SB_NB8768284 done\n",
      "\n",
      "====================\n",
      "\n",
      "20.0863261222839 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "4.16666666666667%\n",
      "\n",
      "sample WSSS_A_KID9369817\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample WSSS_A_KID9369817 done\n",
      "\n",
      "====================\n",
      "\n",
      "58.5618648529053 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "6.25%\n",
      "\n",
      "sample WSSS_A_KID9369818\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample WSSS_A_KID9369818 done\n",
      "\n",
      "====================\n",
      "\n",
      "53.1012592315674 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "8.33333333333333%\n",
      "\n",
      "sample WSSS_A_KID9369819\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample WSSS_A_KID9369819 done\n",
      "\n",
      "====================\n",
      "\n",
      "59.6246421337128 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "10.4166666666667%\n",
      "\n",
      "sample WSSS_A_KID9369820\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample WSSS_A_KID9369820 done\n",
      "\n",
      "====================\n",
      "\n",
      "58.8093395233154 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "12.5%\n",
      "\n",
      "sample WSSS_A_KID9369821\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample WSSS_A_KID9369821 done\n",
      "\n",
      "====================\n",
      "\n",
      "1.93324868679047 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "14.5833333333333%\n",
      "\n",
      "sample WSSS_A_KID9383375\n",
      "\n",
      "reading in adata\n",
      "\n",
      "sample Well_KID10715496 done\n",
      "\n",
      "====================\n",
      "\n",
      "47.8265991210938 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "66.6666666666667%\n",
      "\n",
      "sample Well_KID10715497\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample Well_KID10715497 done\n",
      "\n",
      "====================\n",
      "\n",
      "27.4670579433441 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "68.75%\n",
      "\n",
      "sample Well_KID10715498\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample Well_KID10715498 done\n",
      "\n",
      "====================\n",
      "\n",
      "23.3502876758575 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "70.8333333333333%\n",
      "\n",
      "sample KID_CZI13249517\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249517 done\n",
      "\n",
      "====================\n",
      "\n",
      "1.17909582853317 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "72.9166666666667%\n",
      "\n",
      "sample KID_CZI13249518\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249518 done\n",
      "\n",
      "====================\n",
      "\n",
      "4.98729538917542 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "75%\n",
      "\n",
      "sample KID_CZI13249519\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249519 done\n",
      "\n",
      "====================\n",
      "\n",
      "1.0351088086764 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "77.0833333333333%\n",
      "\n",
      "sample KID_CZI13249520\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249520 done\n",
      "\n",
      "====================\n",
      "\n",
      "4.47489881515503 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "79.1666666666667%\n",
      "\n",
      "sample KID_CZI13249521\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249521 done\n",
      "\n",
      "====================\n",
      "\n",
      "47.7436892986298 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "81.25%\n",
      "\n",
      "sample KID_CZI13249522\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249522 done\n",
      "\n",
      "====================\n",
      "\n",
      "1.20854376157125 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "83.3333333333333%\n",
      "\n",
      "sample KID_CZI13249523\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249523 done\n",
      "\n",
      "====================\n",
      "\n",
      "55.3610756397247 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "85.4166666666667%\n",
      "\n",
      "sample KID_CZI13249524\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249524 done\n",
      "\n",
      "====================\n",
      "\n",
      "1.16308966875076 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "87.5%\n",
      "\n",
      "sample KID_CZI13249525\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249525 done\n",
      "\n",
      "====================\n",
      "\n",
      "41.0355114936829 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "89.5833333333333%\n",
      "\n",
      "sample KID_CZI13249527\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249527 done\n",
      "\n",
      "====================\n",
      "\n",
      "1.5799675265948 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "91.6666666666667%\n",
      "\n",
      "sample KID_CZI13249528\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249528 done\n",
      "\n",
      "====================\n",
      "\n",
      "1.42685612042745 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "93.75%\n",
      "\n",
      "sample KID_CZI13249529\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249529 done\n",
      "\n",
      "====================\n",
      "\n",
      "58.7329061031342 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "95.8333333333333%\n",
      "\n",
      "sample KID_CZI13249530\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249530 done\n",
      "\n",
      "====================\n",
      "\n",
      "1.26886553764343 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "97.9166666666667%\n",
      "\n",
      "sample KID_CZI13249531\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249531 done\n",
      "\n",
      "====================\n",
      "\n",
      "58.5031354427338 seconds\n",
      "\n",
      "***====================***\n",
      "\n",
      "***====================***\n",
      "\n",
      "100%\n",
      "\n",
      "sample KID_CZI13249532\n",
      "\n",
      "reading in adata\n",
      "\n",
      "adding genotyping assignments\n",
      "\n",
      "adding metadata\n",
      "\n",
      "adding demographics\n",
      "\n",
      "adding velocyto\n",
      "\n",
      "sample KID_CZI13249532 done\n",
      "\n",
      "====================\n",
      "\n",
      "47.6271138191223 seconds\n",
      "\n",
      "***====================***\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#now read it all in\n",
    "dir <- \"/lustre/scratch126/cellgen/team297/bs16/current_projects/kidney_glomTI_response/data/seq_data/10X_3p\"\n",
    "adata_list <- lapply(1:nrow(fn_in), function(x){\n",
    "    message(\"***====================***\")\n",
    "    message(paste0(x/nrow(fn_in)*100, \"%\"))\n",
    "    start_time <- Sys.time()\n",
    "    f <- fn_in[x, \"filename\"]\n",
    "    ch <- fn_in[x, \"channel\"]\n",
    "    message(\"sample \", ch)\n",
    "#read in adata\n",
    "    message(\"reading in adata\")\n",
    "    #adata <- afh$anndata_from_h5(file.path(dir, f, \"cellbender_corrected_filtered.h5\"))   \n",
    "    #adata$var_names_make_unique()\n",
    "    #orig_adata <- sc$read_10x_h5(file.path(dir, f,  \"filtered_feature_bc_matrix.h5\" ))\n",
    "    adata <- sc$read_h5ad(file.path(dir, f, \"scAR_denoised.h5ad\"))\n",
    "    adata$var_names_make_unique()\n",
    "    adata$obs$barcode <- adata$obs_names$values\n",
    "    adata$obs$channel <- ch #add the channel name\n",
    "   #make the denoised counts a sparse layer.\n",
    "    adata$layers = list(\"denoised_counts\" = sp$sparse$csr_matrix(adata$layers['denoised_counts']),\n",
    "                        \"raw_counts\" = adata$layers['raw_counts']\n",
    "                       ) #this will help with memory probably\n",
    "#add genotyping assignments\n",
    "    message(\"adding genotyping assignments\")\n",
    "    assignments <- cell_assignments[cell_assignments$channel %in% ch, c(2,3, 5,6)] #these are ordered as per the barcode file\n",
    "    adata$obs <- cbind(adata$obs, assignments)\n",
    "    \n",
    "#sort out the metadata\n",
    "        message(\"adding metadata\")\n",
    "        metadata = sequencing_inventory\n",
    "        rownames(metadata) <- metadata$sanger_ID\n",
    "        adata$obs <- cbind(adata$obs, metadata[adata$obs$channel, ])\n",
    "    \n",
    "    \n",
    "#add in some donor demographics ####    \n",
    "    message(\"adding demographics\")\n",
    "    donor_demographics <- read.csv(\"~/ben_nfs/sample_donor_data/donor_demographics.csv\", row.names = 1)\n",
    "    donor_demographics <- donor_demographics[adata$obs$genotype, ]\n",
    "    adata$obs <- cbind(adata$obs, donor_demographics) #bind this to adata$obs - this will result in NAs but we will remove the doublets in due cours\n",
    "\n",
    "#add the velocyto data\n",
    "    message(\"adding velocyto\")\n",
    "    ldata = scv$read(file.path(dir, f, 'velocyto', paste0(f, '.loom')))\n",
    "    #fix up the loom names\n",
    "    loom_names = gsub(f, \"\", ldata$obs_names$values)\n",
    "    loom_names = gsub(\":\", \"\", loom_names)\n",
    "    loom_names = gsub(\"x\", \"-1\", loom_names)\n",
    "    ldata$obs_names = paste0(loom_names, \"_\",  f)\n",
    "    #merge\n",
    "    adata = scv$utils$merge(adata, ldata)\n",
    "    rm(ldata)\n",
    "    gc() #clean\n",
    "#return the adata\n",
    "    message(\"sample \", ch, \" done\")\n",
    "    message(\"====================\")\n",
    "        #time\n",
    "    end_time <- Sys.time()\n",
    "    time =  end_time - start_time\n",
    "    message(paste(time, 'seconds'))\n",
    "    message(\"***====================***\")\n",
    "\n",
    "return(adata)\n",
    "})\n",
    "adata <- sc$concat(adata_list)\n",
    "vr <- adata_list[[1]]$var\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "realistic-association",
   "metadata": {},
   "outputs": [],
   "source": [
    "#concatenate these data\n",
    "adata <- sc$concat(adata_list)\n",
    "vr <- adata_list[[1]]$var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "essential-heaven",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sort out symbol etc..\n",
    "adata$var$ID <- vr$gene_ids\n",
    "adata$var$Symbol <- rownames(vr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "uniform-nickel",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata$write_h5ad(\"data/raw/10X_3p/healthy_raw_pre_QC.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abstract-nancy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.0.4 (2021-02-15)\n",
       "Platform: x86_64-pc-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.1 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0\n",
       "LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] parallel  stats4    stats     graphics  grDevices utils     datasets \n",
       "[8] methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] paletteer_1.4.1             pbapply_1.7-0              \n",
       " [3] SoupX_1.5.2                 igraph_1.2.6               \n",
       " [5] DropletUtils_1.10.3         pheatmap_1.0.12            \n",
       " [7] cowplot_1.1.1               scater_1.18.6              \n",
       " [9] ggplot2_3.4.1               scran_1.18.5               \n",
       "[11] SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0\n",
       "[13] Biobase_2.50.0              GenomicRanges_1.42.0       \n",
       "[15] GenomeInfoDb_1.26.7         IRanges_2.24.1             \n",
       "[17] S4Vectors_0.28.1            BiocGenerics_0.36.1        \n",
       "[19] MatrixGenerics_1.2.1        matrixStats_0.63.0         \n",
       "[21] reticulate_1.28            \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] uuid_0.1-4                plyr_1.8.8               \n",
       "  [3] repr_1.1.3                lazyeval_0.2.2           \n",
       "  [5] sp_1.6-0                  splines_4.0.4            \n",
       "  [7] BiocParallel_1.24.1       listenv_0.9.0            \n",
       "  [9] scattermore_0.8           digest_0.6.31            \n",
       " [11] htmltools_0.5.4           viridis_0.6.2            \n",
       " [13] fansi_1.0.4               magrittr_2.0.3           \n",
       " [15] tensor_1.5                cluster_2.1.1            \n",
       " [17] ROCR_1.0-11               limma_3.46.0             \n",
       " [19] globals_0.16.2            R.utils_2.12.2           \n",
       " [21] spatstat.sparse_3.0-0     colorspace_2.1-0         \n",
       " [23] ggrepel_0.9.3             dplyr_1.1.0              \n",
       " [25] crayon_1.5.2              RCurl_1.98-1.10          \n",
       " [27] jsonlite_1.8.4            spatstat.data_3.0-0      \n",
       " [29] progressr_0.13.0          survival_3.2-10          \n",
       " [31] zoo_1.8-11                glue_1.6.2               \n",
       " [33] polyclip_1.10-4           gtable_0.3.1             \n",
       " [35] zlibbioc_1.36.0           XVector_0.30.0           \n",
       " [37] leiden_0.3.9              DelayedArray_0.16.3      \n",
       " [39] BiocSingular_1.6.0        Rhdf5lib_1.12.1          \n",
       " [41] future.apply_1.10.0       HDF5Array_1.18.1         \n",
       " [43] abind_1.4-5               scales_1.2.1             \n",
       " [45] edgeR_3.32.1              spatstat.random_3.1-3    \n",
       " [47] miniUI_0.1.1.1            Rcpp_1.0.10              \n",
       " [49] viridisLite_0.4.1         xtable_1.8-4             \n",
       " [51] dqrng_0.3.0               rsvd_1.0.5               \n",
       " [53] htmlwidgets_1.6.1         httr_1.4.4               \n",
       " [55] RColorBrewer_1.1-3        ellipsis_0.3.2.9000      \n",
       " [57] Seurat_4.3.0              ica_1.0-3                \n",
       " [59] pkgconfig_2.0.3           R.methodsS3_1.8.2        \n",
       " [61] scuttle_1.0.4             uwot_0.1.14              \n",
       " [63] deldir_1.0-6              locfit_1.5-9.4           \n",
       " [65] utf8_1.2.3                reshape2_1.4.4           \n",
       " [67] tidyselect_1.2.0          rlang_1.0.6              \n",
       " [69] later_1.3.0               munsell_0.5.0            \n",
       " [71] tools_4.0.4               cli_3.6.0                \n",
       " [73] generics_0.1.3            ggridges_0.5.4           \n",
       " [75] stringr_1.5.0             evaluate_0.20            \n",
       " [77] fastmap_1.1.0             goftest_1.2-3            \n",
       " [79] rematch2_2.1.2            fitdistrplus_1.1-8       \n",
       " [81] purrr_1.0.1               RANN_2.6.1               \n",
       " [83] nlme_3.1-152              future_1.31.0            \n",
       " [85] sparseMatrixStats_1.2.1   mime_0.12                \n",
       " [87] R.oo_1.25.0               compiler_4.0.4           \n",
       " [89] beeswarm_0.3.1            plotly_4.10.1            \n",
       " [91] png_0.1-8                 spatstat.utils_3.0-1     \n",
       " [93] tibble_3.1.8              statmod_1.4.35           \n",
       " [95] stringi_1.7.12            lattice_0.20-41          \n",
       " [97] bluster_1.0.0             IRdisplay_1.0            \n",
       " [99] Matrix_1.5-3              vctrs_0.5.2              \n",
       "[101] pillar_1.8.1              lifecycle_1.0.3          \n",
       "[103] rhdf5filters_1.2.1        spatstat.geom_3.0-6      \n",
       "[105] lmtest_0.9-40             RcppAnnoy_0.0.20         \n",
       "[107] BiocNeighbors_1.8.2       data.table_1.14.6        \n",
       "[109] bitops_1.0-7              irlba_2.3.5.1            \n",
       "[111] httpuv_1.6.8              patchwork_1.1.2          \n",
       "[113] R6_2.5.1                  promises_1.2.0.1         \n",
       "[115] KernSmooth_2.23-18        gridExtra_2.3            \n",
       "[117] vipor_0.4.5               parallelly_1.34.0        \n",
       "[119] codetools_0.2-18          MASS_7.3-53.1            \n",
       "[121] rhdf5_2.34.0              withr_2.5.0              \n",
       "[123] SeuratObject_4.1.3        sctransform_0.3.5        \n",
       "[125] GenomeInfoDbData_1.2.4    grid_4.0.4               \n",
       "[127] beachmat_2.6.4            IRkernel_1.1.1           \n",
       "[129] tidyr_1.3.0               DelayedMatrixStats_1.12.3\n",
       "[131] Rtsne_0.16                spatstat.explore_3.0-6   \n",
       "[133] pbdZMQ_0.3-5              getPass_0.2-2            \n",
       "[135] shiny_1.7.4               base64enc_0.1-3          \n",
       "[137] ggbeeswarm_0.6.0         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
